#4 feature(View): Выполнена лабораторная работа №4 по ООАиПр

Melaved opened the pull request.

Melaved committed:
feature(View): реализован ContactControl

- реализован пользовательский элемент управления ContactControl, являющийся правой панелью главного окна(кнопка Apply перенесена в ContactControl)
- реализован DataTemplate для списка контактов. Теперь список контактов отображает не только имя контакта, но и его номер телефона
- DataTemplate вынесен в Windows.Resources

Melaved committed:
feature(View): Реализована валидация для вводимых данных

- реализована валидация на основе интерфейса INotifyDataErrorInfo для вводимых пользователем данных
- реализовано преформатирование вводимого текста для поля Phone Number, запрещающее ввод любых недопустимых символов

inittose commented on View/Controls/ContactControl.xaml:
```diff
@@ -0,0 +1,112 @@
+                    <Setter Property="ToolTip" Value="{Binding RelativeSource={RelativeSource Self}, Path=(Validation.Errors)[0].ErrorContent}"/>
+                </Trigger>
+            </Style.Triggers>
+        </Style>
```
Сейчас ошибка подсвечивается не как в макете. Сделать фон ошибки цветом `#ffdddd`. Чтобы убрать цвет border при ошибке, нужно прописать в стилях TextBox:
```xaml
<Setter Property="Validation.ErrorTemplate" Value="{x:Null}"/>
```

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,173 @@ public Contact SelectedContact
 
         /// <summary>
-        /// Возвращает и задаетт значение, указывающее, видна ли кнопка "Применить".
+        /// Возврашает и задает контакт, который редактируется в данный момент.
```
*Возвращает

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,173 @@ public Contact SelectedContact
+                    }
+
+                    OnPropertyChanged(nameof(IsContactSelected));
                 }
```
Вложенность от которой можно избавиться. Мы убираем вложенность, чтобы уменьшить нагромождение кода и повысить читаемость кода 
Проверить во всем решении

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,173 @@ public Contact SelectedContact
+        /// <returns>True, если можно применить изменения.</returns>
+        private bool CanApplyChanges(object parameter) => IsEditingContact 
+                                                          && EditingContact != null 
+                                                          && !EditingContact.HasErrors;
```
Лишние пробелы в конце строк. Не стоит добавлять пробелы, когда переносишь часть строки, лучше заменить на:
```csharp
        private bool CanApplyChanges(object parameter) =>
            IsEditingContact && EditingContact != null && !EditingContact.HasErrors;
```
или:
```csharp
        private bool CanApplyChanges(object parameter) =>
            IsEditingContact
            && EditingContact != null
            && !EditingContact.HasErrors;
```

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <param name="e">Аргументы события, содержащие введённый текст.</param>
+        private void PhoneNumber_PreviewTextInput(object sender, TextCompositionEventArgs e)
+        {
+            var regex = new Regex(@"^[0-9+() -]*$");
```
Данное регулярное выражение (regex) относится к Contact, поэтому создай в классе Contact константу (или статическое свойство) типа `Regex` и используй её

inittose commented on View/Model/Contact.cs:
```diff
@@ -72,30 +112,117 @@ public string PhoneNumber
+        /// </summary>
+        private void ValidateName()
+        {
+            if (Name.Length > 100)
```
100 - магическое число, т.е. числовое значение, смысл которого может быть неочевиден. Это считается плохой практикой еще потому, что если придется менять логику (в данном случае верхнюю границу длины имени), то это придется менять везде, поэтому замени его на константу (например, `MaxNameLength`)
По аналогии сделай константы для других свойств

inittose commented on View/Model/Contact.cs:
```diff
@@ -72,30 +112,117 @@ public string PhoneNumber
+        /// </summary>
+        private void ValidatePhoneNumber()
+        {
+            if (PhoneNumber.Length > 100
+                || !Regex.IsMatch(PhoneNumber, "^[0-9+() -]*$"))
```
Используй более сложное регулярное выражение, есть много сайтов с уже готовыми regex, например [этот](https://ihateregex.io/expr/phone/)

Melaved replied:
С этим Regex проблема с частичным вводом.
Можно конечно сделать два поля Regex, для PhoneNumber_PreviewTextInput использовать простой Regex,
а для ValidatePhoneNumber более сложный(я пока сделал так, но если надо поменяю обратно)

inittose commented on View/Model/Contact.cs:
```diff
@@ -72,30 +112,117 @@ public string PhoneNumber
+        /// </summary>
+        private void ValidateEmail()
+        {
+            if (Email.Length > 100 || !Email.Contains("@"))
+            {
+                _errors[nameof(Email)] = "Электронная почта должна содержать"
+                                         + " '@' и не превышать 100 символов.";
+            }
```
Используй регулярное выражение для [электронной почты](https://ihateregex.io/expr/email/)

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    /// <summary>
+    /// Логика взаимодействия для ContactControl.xaml
+    /// Контрол для отображения и ввода контактной информации.
```
Слово "Контрол" лучше заменить на "Элемент управления"

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <param name="e">Аргументы события, содержащие вставляемый текст.</param>
+        private void PhoneNumber_Pasting(object sender, DataObjectPastingEventArgs e)
+        {
+            if (e.DataObject.GetDataPresent(typeof(string)))
+            {
+                var text = (string)e.DataObject.GetData(typeof(string));
+
+                var regex = new Regex(@"^[0-9+() -]*$");
+
+                if (!regex.IsMatch(text))
+                {
+                    e.CancelCommand();
+                }
+            }
+            else
+            {
+                e.CancelCommand();
+            }
+        }
```
Убрать лишнюю вложенность

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            if (e.DataObject.GetDataPresent(typeof(string)))
+            {
+                var text = (string)e.DataObject.GetData(typeof(string));
+
+                var regex = new Regex(@"^[0-9+() -]*$");
+
+                if (!regex.IsMatch(text))
```
Лишнии пустые строки
После объявления локальной переменной пустая строка не ставится, проверить во всем решении

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,63 @@
+        private void PhoneNumber_PreviewTextInput(object sender, TextCompositionEventArgs e)
+        {
+            var regex = new Regex(@"^[0-9+() -]*$");
+
```
Лишняя пустая строка

inittose replied:
Не исправлено

Melaved replied:
этой строки больше нет, но если я правильно понимаю где убрать, то  убрал пустую строку после строки
`var newText = textBox.Text.Insert(textBox.CaretIndex, e.Text);`
![image](https://github.com/user-attachments/assets/d23681f0-c805-4586-8e4a-046e1df38139)

Melaved committed:
feature(View): Исправил замечания

- убрал лишнюю вложенность
- убрал пустые строки
- вынес значения максимальных длин полей в константы
- вынес  регулярное выражение и сделал его статическим
- теперь программа имеет для регулярных выражения для номера телефона, одно для частичного ввода, второе(сложное) для полной проверки

inittose commented on View/Model/Contact.cs:
```diff
@@ -1,12 +1,40 @@
+        /// <summary>
+        /// Регулярное выражение для проверки номера телефона.
+        /// </summary>
+        public static readonly Regex SimplePhoneNumberRegex = new Regex(@"^[0-9+() -]*$");
+        public static readonly Regex StrictPhoneNumberRegex = new Regex(@"^\+?(\d{1,3})?[-. (]*(\d{1,4})[-. )]*(\d{1,4})[-. ]*(\d{1,9})$");
```
`SimplePhoneNumberRegex` переименовать в `PhoneNumberMask`
`StrictPhoneNumberRegex ` переименовать в `PhoneNumberRegex `, добавить комментарий к нему и перенести new Regex на новую строку (сейчас строка более 100 символов)
Также добавить поле с регулярным выражением для Email

inittose replied:
Часть строки `new Regex` перенести полностью, где длина строки более 100 символов, вот так:
```csharp
public static readonly Regex EmailRegex = 
    new Regex(@"^[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+$");
```

inittose commented on View/Model/Contact.cs:
```diff
@@ -72,25 +132,114 @@ public string PhoneNumber
+            if (PhoneNumber.Length > MaxPhoneNumberLength
+                || !Contact.StrictPhoneNumberRegex.IsMatch(PhoneNumber))
+            {
+                _errors[nameof(PhoneNumber)] = $"Номер телефона должен быть в формате '+7 (999) 123-45-67' "
+                                               + $"и не превышать {MaxPhoneNumberLength} символов.";
```
Строки длиной более 100 символов
Убрать большой отступ

inittose replied:
Под словом отступ я имею ввиду часть кода, которая перенесена на новую строку. использовать отступ, который ставит IDE по умолчанию. Т. е. выглядеть должно так:
```csharp
                _errors[nameof(PhoneNumber)] = $"Номер телефона должен быть в формате " +
                    $"'+7 (999) 123-45-67' и не превышать " +
                    $"{MaxPhoneNumberLength} символов.";
```
Или так:
```csharp
                _errors[nameof(PhoneNumber)] = 
                    $"Номер телефона должен быть в формате " +
                    $"'+7 (999) 123-45-67' и не превышать " +
                    $"{MaxPhoneNumberLength} символов.";
```

inittose commented on View/Model/Contact.cs:
```diff
@@ -72,25 +132,114 @@ public string PhoneNumber
+            if (Email.Length > MaxEmailLength
+                || !Regex.IsMatch(Email, @"[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+"))
+            {
+                _errors[nameof(Email)] = $"Электронная почта должна содержать '@', "
+                                         + $"иметь корректный формат и не превышать "
+                                         + $"{MaxEmailLength} символов.";
```
Убрать большой отступ

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,169 @@ public Contact SelectedContact
 
                 _selectedContact = value;
                 OnPropertyChanged(nameof(SelectedContact));
+
```
Лишняя пустая строка

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,169 @@ public Contact SelectedContact
         {
-            if (SelectedContact == null)
+            if (SelectedContact != null)
             {
-                return;
+                EditingContact = SelectedContact.Clone();
+                EditingContact.IsEditing = true;
+                EditingContact.IsNewContact = false;
+                OnPropertyChanged(nameof(EditingContact));
             }
```
Уйти от вложенности

inittose commented on View/ViewModel/MainVM.cs:
```diff
@@ -57,207 +41,169 @@ public Contact SelectedContact
+                Contacts.Remove(SelectedContact);
+                ContactSerializer.SaveContacts(Contacts);
+                SelectedContact = Contacts.Count > 0 ? Contacts[0] : null;
             }
```
Уйти от вложенности

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -27,9 +28,10 @@ public ContactControl()
+            var newText = textBox.Text.Insert(textBox.CaretIndex, e.Text);
 
-            if (!regex.IsMatch(e.Text))
+            if (!Contact.StrictPhoneNumberRegex.IsMatch(newText))
```
Здесь и в следующем методе надо использовать маску, иначе не получится ничего ввести

inittose commented on View/Controls/ContactControl.xaml.cs:
```diff
@@ -0,0 +1,61 @@
+﻿using System.Text.RegularExpressions;
```
Не используется

Melaved committed:
style(View): Исправил замечания

Melaved committed:
style(View): Исправил замечания

inittose commented on View/Controls/ContactControl.xaml:
```diff
@@ -0,0 +1,113 @@
+            VerticalAlignment="Top"
+            Command="{Binding DataContext.ApplyCommand, RelativeSource={RelativeSource AncestorType=Window}}"
+            Content="Apply"
+            IsEnabled="{Binding DataContext.CanApply, RelativeSource={RelativeSource AncestorType=Window}}"
```
В MainVM нет свойства CanApply
Кнопка Apply не работает при успешной валидации:
![изображение](https://github.com/user-attachments/assets/bbd6733f-2fcf-40e3-a24f-23ceeffb6bb4)

inittose commented on View/Controls/ContactControl.xaml:
```diff
@@ -0,0 +1,113 @@
+            Command="{Binding DataContext.ApplyCommand, RelativeSource={RelativeSource AncestorType=Window}}"
+            Content="Apply"
+            IsEnabled="{Binding DataContext.CanApply, RelativeSource={RelativeSource AncestorType=Window}}"
+            Visibility="{Binding DataContext.IsEditingContact, RelativeSource={RelativeSource AncestorType=Window}, Converter={StaticResource BooleanToVisibilityConverter}}"/>
```
При открытии программы видна кнопка Apply, а textbox'ы не readonly:
![изображение](https://github.com/user-attachments/assets/a26b5741-0e6a-4272-a8f2-ceea38cafc08)
IsEditingContact не может прибиндиться, потому что свойство private

Melaved committed:
feature(View): Исправил замечания
